### ISSUE TITLE
Forward-compatibility for special case APIs

### ISSUE THREAD

**Issue by zooba** [capi-workgroup/api-revolution#4]

**Comment by zooba** [comment]
> Given that you can compile wheels for the new & old versions separately ...

Uhm... are you suggesting that people should just compile wheels for each version? What's the point of having a stable ABI/API between versions then?

**Comment by encukou** [comment]
Yeah, you're right. It expands the scope of the stable ABI, but it'd be a welcome enhancement.

https://github.com/capi-workgroup/api-evolution/issues/1 is another possible approach to solving it.

**Comment by zooba** [comment]
The two approaches come at it from different directions.

This one gives us a path to design an ABI level that remains compatible over time, and then we can build helpers on top of it to make life easier for end users (similar in principle to #1 Native Interface proposal, but this would work as a component of that rather than an alternative).

The other approach assumes that we'll continue to add, change and remove ABI members over time in incompatible ways, but will include shims in user's code as needed to handle the differences. This doesn't handle forward-compatibility, because we still can't predict future changes, but it does allow us to handle backwards compatibility (though, IMHO, no better than this proposal). It also doesn't allow us to safely provide API-level optimisations or an interface that can be efficient for other Python implementations.

But as notable from being in separate repositories, they aren't mutually exclusive. We can invest in a shim library (probably HPy is the right place for it?) while also developing an API structure that can eventually be the new stable ABI. If anyone does think they are totally exclusive, I'd like the opportunity to sell this one better, because I don't believe they are.

**Comment by zooba** [comment]
Oh, we also don't necessarily have to define every slots struct as belonging to the stable ABI if the API to get them does.

It would probably be appreciated by users if we at least have baseline functionality guaranteed to not go away, but provided that we can always point at what the fallback interface ought to be, we can always remove[^2] more specific slots in a later version. Code will still work, which is the promise, even if it loses a bit of performance.[^1]

[^1]: And yes, I know I suggested earlier that an extension might do something other than call the fallback functionality. But we're all consenting adults here. If they do that, they are intentionally planning to break themselves.

[^2]: Remove, but not modify.

**Comment by encukou** [comment]
The more I think about this the more questoins I have :)

We probably want a linear sequence of fallbacks.
Let's say we add a context argument to all the signatures. Is the proper order
`PY_SLOT_GETITEM_INT_CTX` -> `PY_SLOT_GETITEM_INT` -> `PY_SLOT_GETITEM_CTX` -> `PY_SLOT_GETITEM`, or the other reasonable permutation?

What happens if a user uses the wrong order? A subtle behaviour difference? Is that a bug in the exporting object (do we require that the fallbacks have equivalent behaviour)?
What if different slots are defined in different classes in the MRO?

I can see fallbacks being optional: If there's a slot to "get an int *value* from a dict", but it overflows, you might, but might not, want to fall back to "get an object from dict".

**Comment by zooba** [comment]
The "proper order" will depend on the particular slots, so it'd have to be part of the definition. For example, `PY_SLOT_GETITEM_CTX` would probably be documented as "if unavailable, `PY_SLOT_GETITEM` is the best fallback, except for <difference in performance/behaviour>".

If the user uses the wrong order, then one of the earlier checks will succeed and none of the later ones will occur. So if they check for `PY_SLOT_GETITEM` first, when that succeeds (presumably always), every `else` condition after it is essentially dead code. (Unless we one day deprecate that particular slot, in which case they can get warnings for a while and _then_ the next `else` condition will be taken. Again, "getitem(object)" is an unlikely candidate for removal, but the principle applies to all slots.)

> What if different slots are defined in different classes in the MRO?

The identifier is the name, so `PY_SLOT_GETITEM` will be fully provided by one implementation (presumably the first in the MRO), and would include all members that are part of it (we might put `contains` and `getitem` in the same struct, for example). 

I haven't specified any MRO logic at all here, btw, since we don't currently have that in native types. Most likely, your `PyTypeObject` would specify exactly one handler and then that would manually delegate unrecognised names to another type's handler. User-defined classes would probably have a generic implementation for a slot that does the attribute (or legacy slot) lookup when invoked - these slots are about finding the right native function to call, and calling `PyObject_GetItem` on an object that doesn't support getitem is technically okay, even though it's going to raise an error.

> I can see fallbacks being optional: If there's a slot to "get an int _value_ from a dict", but it overflows, you might, but might not, want to fall back to "get an object from dict".

Yeah, exactly. Or even just getting an int value from a long-like object. _Any_ native type could implement a `PY_SLOT_AS_INT64` slot with their own `int64_t (*get_value)(PySlots*)` member, whereas today we would need the type to create an actual PyLongObject so that our conversion functions work.

And "this should've worked, but can't" is often enough of an error condition to make it an error immediately rather than going through and trying more native APIs.

**Comment by encukou** [comment]
It seems that one way to look at this is as an extension of the current [sub-slot structs](https://docs.python.org/3/c-api/typeobj.html?highlight=pytypeobject#sub-slots), but with an extra layer of indirection (which the object controls).
The indirection might be too slow for core (and thus Cython); should there be an alternate API that's faster for common operations?

It would be helpful to have a concrete example for how the “provider” side would be implemented -- what a type author should do to make `PyObject_GetSlots` work for a particular class.

**Comment by zooba** [comment]
In the simplest case, the implementer would do something like this:

```c
// We've defined the interface somewhere in our public headers
struct PySlot_GetSetDelItem {
    struct PySlots_Base base;
    PyObject * (*getitem)(PySlots *self, PyObject *key);
    PyObject * (*setitem)(PySlots *self, PyObject *key, PyObject *value);
    PyObject * (*delitem)(PySlots *self, PyObject *key);
};

// The type implementer defines a (potentially static) function table in their code
struct PySlots_GetItem _dict_getitem_slots = {
    _PyDictSlots_Release,
    &_PyDictSlots_GetItem,
    &_PyDictSlots_SetItem,
    &_PyDictSlots_DelItem
};

// Assuming we've defined PySlots as:
struct PySlots {
    void *handle;
    PySlotName name;
    struct PySlots_Base *slots;
};

// Their "...Slots_Get" function returns the right slots for the requested name.
// This function is found in the PyTypeObject of 'o'
static int
_PyDictSlots_Get(PyObject *o, PySlotName name, PySlots *slots)
{
    switch (name) {
    case PY_SLOT_GETSETDELITEM:
        slots->handle = Py_NewRef(o);
        slots->name = name;     // I just added this, realised we need it for Release
        slots->slots = (struct PySlots_Base*)&_dict_getitem_slots;
        return 0;
    }
    return _PySlots_GenericGet(o, name, slots);
}

static int
_PyDictSlots_Release(PySlots *slots)
{
    switch (slots->name) {
    case Py_SLOT_GETSETDELITEM:
        Py_DECREF((PyObject *)slots->handle);
        return 0;
    }
    return _PySlots_GenericRelease(slots);
}

// For this example we just redirect to the existing function, but this could
// contain the actual implementation. It doesn't necessarily have to just be
// a wrapper around the more obvious function.
static int
_PyDictSlots_GetItem(PySlots *self, PyObject *key)
{
    return PyDict_GetItem((PyObject *)self->handle, key);
}
```

I also wrote up a rough example of an efficient "AsLong" slots, but posted it [in a Gist](https://gist.github.com/zooba/3e936c932dedea4a106cfee96e6e78ed) rather than cluttering up this thread more. It's a very hypothetical idea, but I think it shows the potential efficiency gains more than GetItem does.

**Comment by encukou** [comment]
Thanks!
I still think the indirection will be problematic for core CPython and Cython. We might need something like baking “current” slots into the type object more statically (and with lower compatibility expectations). But that sounds doable.

Two questions to hash out:
- What happens if I override `__getitem__` in a subclass?
- What happens if I replace `__getitem__` from Python on a mutable class that exposes slots?

Currently, Python knows about all slots so it can update them appropriately.

**Comment by zooba** [comment]
> * What happens if I override `__getitem__` in a subclass?

Then you'll also override `_PyDictSlots_Get` as your `tp_slots` or whatever, and will return something other than `_dict_getitem_slots ` when queried (though you might choose to be compatible/brave enough to reuse some of its functions).

> * What happens if I replace `__getitem__` from Python on a mutable class that exposes slots?

I imagine for a mutable class, the slots would be generic (i.e. they'd do the `return getattr(self, "__getitem__")(key)`). But while a generic `GetSlots` would return these generic handlers, a specialising `GetSlots` could use caching/profiling/checks to return faster ones where possible.

This API doesn't really do a huge amount out of the box for Python user-defined classes. But in any case, those are based on our C type, and so we still know all the slots. Just that our `tp_mapping` (or equivalent) structs of callables would be a completely private cache.

> I still think the indirection will be problematic for core CPython and Cython.

You mean performance-wise? Obviously there'll be a minor perf regression compared to assuming the type and calling the function directly, but by using the indirection it means core native functions can efficiently use 3rd party types, whereas today that only works when they haven't overridden our own implementations.

I'm sure Cython will choose to make assumptions about internal layouts that will eventually come back to bite them (after all, that's exactly what they do today). e.g. they'll probably assume that the getitem slots for one dict will work for any dict, which would probably be true up until the point where it isn't.

What they'd gain is the option to have unconditional code for using faster interfaces that still works (slower) on older versions. So if we add a `PySLOT_FAST_MUL_INT` to multiply a PyLongObject by an int constant, they can start querying for that on all versions (with a fallback to regular `__mul__` for versions that don't have it yet, which must already have existed). So all Cythonised extensions could be compiled once to cover all active CPython releases and still be efficient, rather than users having to build one wheel per version.

**Comment by encukou** [comment]
Let me give concrete hypothetical scenarios. These involve inheritance, and exploit (perhaps too much) the mechanism that decouples the  slots from the interpreter.
I don't see a good way to make them both work. If you do, or you see something that shouldn't be possible, could you correct me?

## Scenario 1

- `foo_spam`, a third-party library, exports `SuperDataClass`. It uses `PY_SLOT_REPR` for its `__repr__()` implementation.
- `foo_ide`, my editor, uses `PY_SLOT_REPR` to display info in the debugger.
- In my code, I subclass that to make `HyperDataClass`. I override `PY_SLOT_REPR` as I need to tweak it a bit.
- CPython 3.42 introduces `PY_SLOT_REPR_UTF8ANDLEN` to allow avoiding `PyUnicode` overhead in `__repr__`.
- `foo_spam.SuperDataClass` and `foo_ide` update to use `PY_SLOT_REPR_UTF8ANDLEN`. Thanks to the magic of slots, they can use it on all Python versions, not just 3.42+

Here, `foo_ide` should see the repr for `HyperDataClass`, not from the superclass. In other words, `PY_SLOT_REPR_UTF8ANDLEN` should not be inherited.

## Scenario 2

- `foo_spam` also defines `SuperArray`, which relies on `NPY_SLOT_NDARRAY` to work as a generalized NumPy array. (This is a slot that the scientific community reserved for NumPy &c, and like the `@` operator, that was a big win in those circles. Unlike new syntax, it can even be used in older CPython versions.)
- My code derives a `HyperArray` which, again, tweaks `PY_SLOT_REPR` a bit.

Here, `NPY_SLOT_NDARRAY` should be inherited from `SuperArray`, so that it still works like the supertype does.

**Comment by zooba** [comment]
Scenario 2 is easy, because your `SuperArray.tp_getslot` will handle the slots it knows and then call `SuperArray.tp_base.tp_getslot` for anything else (or along those lines).

Scenario 1 is hard in general. If a subclass doesn't override a protocol, the superclass provides it, regardless of whether we're talking about this proposal or existing Python code.

Just thinking out loud here, but maybe we could mark some slots as uninheritable (or alternatively, "must be implemented by the most derived class")? If they're ints, maybe the topmost bit is set. So if `PY_SLOT_REPR_UTF8ANDLEN` is marked as such, your subclass will refuse to return `PY_SLOT_REPR_UTF8ANDLEN` even though it doesn't know it even exists, and the caller will call again with `PY_SLOT_REPR`. When you learn it exists, you either implement it directly or by calling your superclass.

**Comment by antocuni** [comment]
My first thought is that this proposal conflates two different goals:
  - ability to be forward _and_ backward compatible with the ABI
  - ability to provide fast/specialized protocols for certain behaviors

It _might_ be that the best way to achieve those goals is to use two different mechanisms. I'm not saying we should and it might be that this is actually the best way (I haven't fully thought about it yet), but I think it's a good idea to be explicit in saying that.

For more context, at some point we thought about having a similar concept in HPy, to solve the "specialized protocol" problem. This idea come from the very first hpy meeting when a lot of details were not designed yet, and we never fully explored it (the fact that it lives in a file named `xxx_unsorted_notes.txt` should tell you something :)).
Nevertheless, this is how it looked like:
https://github.com/hpyproject/hpy/blob/b57a33c1cec766a1cc3e89f6fd1e2eff73ba9381/docs/xxx-unsorted-notes.txt#L36-L58

**Comment by zooba** [comment]
> It _might_ be that the best way to achieve those goals is to use two different mechanisms

Agreed that it's doing two things, but also, we're in a position as the reference implementation to merge stuff like this. I would fully expect anything that layers on top of the CPython API to have to use different mechanisms for different intents, because they have to follow what the API available is (e.g. Cython using clever tricks to provide their fast access for their own types, since there's no CPython API for it). We aren't so restricted, and so "simpler is better" (where simpler == less API surface) is more likely to apply.

I think it's also pretty clear that most of the ABI incompatibilities we've seen/caused have been from providing fast/specialized protocols, with the biggest exception being the effort to fix borrowed references. So from that POV, the ABI churn we're fixing _is_ the desire for specialized protocols.
